File: rust/race-condition/src/race_condition.rs
use std::sync::Arc;
use std::thread;

//----------------------------------Szenario------------------------------------------
/*
 * Dieses Programm simuliert eine Race Condition, indem mehrere Threads
 * gleichzeitig auf eine gemeinsame Variable (`counter`) zugreifen und diese
 * inkrementieren. Der Zugriff erfolgt ohne Synchronisationsmechanismen, wodurch
 * Dateninkonsistenzen entstehen können.
 */
//----------------------------------Ursache-------------------------------------------
/*
 * Jeder Thread greift unsicher (über `unsafe`) auf den `counter` zu und
 * inkrementiert ihn. Es gibt keine Synchronisation, die sicherstellt,
 * dass der Zugriff atomar erfolgt.
 */
//----------------------------------Ergebnis------------------------------------------
/*
 * Aufgrund der Race Condition erreicht der `counter` möglicherweise
 * nicht den erwarteten Wert (`10000`), sondern einen inkonsistenten Wert,
 * da Threads gleichzeitig auf dieselbe Speicheradresse zugreifen.
 */

use std::cell::RefCell;
use std::sync::Arc;
use std::thread;

fn main() {
    // Erstellt einen gemeinsamen Zähler mit `RefCell`
    // Problem: `RefCell` ist nicht threadsicher.
    let counter = Arc::new(RefCell::new(0));

    // Vektor zur Speicherung der Thread-Handles
    let mut handles = vec![];

    for _ in 0..10 {
        // Erstellt eine neue Referenz , die in den Thread übernommen wird
        let counter = Arc::clone(&counter);

        // Erstellt und startet einen neuen Thread
        let handle = thread::spawn(move || {
            for _ in 0..1000 {
                // Race Condition!
                // Mehrere Threads greifen gleichzeitig auf `counter` zu, ohne Synchronisation.
                *counter.borrow_mut() += 1;
            }
        });
        // Speichert das Handle des Threads, um später auf seinen Abschluss zu warten
        handles.push(handle);
    }
    // Wartet darauf, dass alle Threads beendet sind
    for handle in handles {
        handle.join().unwrap();
    }

    // Erwarteter Wert: 10000 (10 Threads * 1000 Inkremente pro Thread)
    println!("Erwarteter Zähler: 10000");

    // Tatsächlicher Wert kann abweichen, da eine Race Condition vorliegt
    println!("Tatsächlicher Zähler: {}", counter.borrow());
}

File: cpp/src/race-condition/race_condition.cpp
#include <chrono>
#include <cstdlib>
#include <iostream>
#include <thread>
#include <vector>

int counter = 0;

// Funktion, die den Zähler inkrementiert
void increment_counter()
{
    for (int i = 0; i < 1000; ++i) {
        ++counter; // Kritischer Abschnitt
        std::this_thread::sleep_for(
            std::chrono::microseconds(rand() % 10)); // Zufällige Verzögerung
    }
}

void simulate_race_condition()
{
    std::vector<std::thread> threads; // Container für die Threads

    // 10 Threads erzeugen und starten
    for (int i = 0; i < 10; ++i) {
        threads.emplace_back(increment_counter); // Threads hinzufügen
    }

    // Warten, bis alle Threads beendet sind
    for (auto& thread : threads) {
        thread.join(); // Join auf jeden Thread im Vector
    }

    // Ergebnis ausgeben
    std::cout << "Erwarteter Zähler: 10000\n"; // 10 Threads x 1000 Inkremente
    std::cout << "Tatsächlicher Zähler: " << counter << std::endl;
}

int main() { simulate_race_condition(); }